src/test_typing.py

```diff
@@ -1516,6 +1575,62 @@ def foo(a: Tuple['T']):
         self.assertEqual(get_type_hints(foo, globals(), locals()),
                          {'a': Tuple[T]})
 
+    def test_forward_recursion_actually(self):
+        def namespace1():
+            a = typing._ForwardRef('A')
+            A = a
+            def fun(x: a): pass
+
+            ret = get_type_hints(fun, globals(), locals())
+            return a
+
+        def namespace2():
+            a = typing._ForwardRef('A')
+            A = a
+            def fun(x: a): pass
+
+            ret = get_type_hints(fun, globals(), locals())
+            return a
+
+        def cmp(o1, o2):
+            return o1 == o2
+
+        r1 = namespace1()
+        r2 = namespace2()
+        self.assertIsNot(r1, r2)
+        self.assertRaises(RecursionError, cmp, r1, r2)
```

Comment:
So this fails in 3.4 because RecursionError doesn't exist there -- recursion raises RuntimeError. You can do something like
```
try:
    exc = RecursionError
except NameError:
    exc = RuntimeError
self.assertRaises(exc, ...)
```

Reply:
Didn't have a chance to look at this till now. Thanks for the snippet :)